drivers/rtc/rtc_ds3231.c

```diff
@@ -0,0 +1,554 @@
+
+#define DS3231_ALARM_2_MINUTES_A2M2    BIT(7)
+#define DS3231_ALARM_2_MINUTES_10      GENMASK(6, 4)
+#define DS3231_ALARM_2_MINUTES_MINUTES GENMASK(3, 0)
+#define DS3231_ALARM_2_HOURS_A2M3      BIT(7)
+#define DS3231_ALARM_2_HOURS_12_24     BIT(6)
+#define DS3231_ALARM_2_HOURS_AM_PM_20  BIT(5)
+#define DS3231_ALARM_2_HOURS_10        BIT(4)
+#define DS3231_ALARM_2_HOURS_HOURS     GENMASK(3, 0)
+#define DS3231_ALARM_2_DAY_DATE_A2M4   BIT(7)
+#define DS3231_ALARM_2_DAY_DATE_DYDT   BIT(6)
+#define DS3231_ALARM_2_DAY_DATE_10     GENMASK(5, 4)
+#define DS3231_ALARM_2_DAY_DATE_MASK   GENMASK(3, 0)
+
+#define DS3231_CONTROL_EOSC  BIT(7)
+#define DS3231_CONTROL_BBSQW BIT(6)
+#define DS3231_CONTROL_CONV  BIT(5)
+#define DS3231_CONTROL_RS2   BIT(4)
+#define DS3231_CONTROL_RS1   BIT(3)
+#define DS3231_CONTROL_INTCN BIT(2)
+#define DS3231_CONTROL_A2IE  BIT(1)
+#define DS3231_CONTROL_A1IE  BIT(0)
+
+#define DS3231_STATUS_OSF     BIT(7)
+#define DS3231_STATUS_EN32KHZ BIT(3)
+#define DS3231_STATUS_BSY     BIT(2)
+#define DS3231_STATUS_A2F     BIT(1)
+#define DS3231_STATUS_A1F     BIT(0)
+
+#define DS3231_AGING_OFFSET_SIGN BIT(7)
+#define DS3231_AGING_OFFSET_DATA GENMASK(6, 0)
+#define DS3231_TEMP_MSB_SIGN     BIT(7)
+#define DS3231_TEMP_MSB_DATA     GENMASK(6, 0)
+#define DS3231_TEMP_LSB_DATA     GENMASK(7, 6)
+
+/* RTC alarm time fields supported by DS3231 */
+#define DS3231_RTC_ALARM_1_TIME_MASK                                                               \
+	(RTC_ALARM_TIME_MASK_SECOND | RTC_ALARM_TIME_MASK_MINUTE | RTC_ALARM_TIME_MASK_HOUR |      \
+	 RTC_ALARM_TIME_MASK_WEEKDAY | RTC_ALARM_TIME_MASK_MONTHDAY)
+
+#define DS3231_RTC_ALARM_2_TIME_MASK                                                               \
+	(RTC_ALARM_TIME_MASK_MINUTE | RTC_ALARM_TIME_MASK_HOUR | RTC_ALARM_TIME_MASK_WEEKDAY |     \
+	 RTC_ALARM_TIME_MASK_MONTHDAY)
+
+/* The DS3231 enumerates months 1 to 12, RTC API uses 0 to 11 */
+#define DS3231_MONTHS_OFFSET 1
+
+/* The DS3231 only supports two-digit years, calculate offset to use */
+#define DS3231_YEARS_OFFSET (2000 - 1900)
+
+/* Macro for interrupt pin code */
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(int1_gpios)
+#define DS3231_INT1_GPIOS_IN_USE 1
+#endif
+
+struct ds3231_config {
+	const struct i2c_dt_spec i2c;
+
+#ifdef DS3231_INT1_GPIOS_IN_USE
+	struct gpio_dt_spec int1;
+#endif /* DS3231_INT1_GPIOS_IN_USE */
+};
+struct ds3231_data {
+	struct k_mutex lock;
+#if DS3231_INT1_GPIOS_IN_USE
+	struct gpio_callback int1_callback;
+	struct k_thread int1_thread;
+	struct k_sem int1_sem;
+
+	K_KERNEL_STACK_MEMBER(int1_stack, CONFIG_RTC_DS3231_THREAD_STACK_SIZE);
+#ifdef CONFIG_RTC_ALARM
+	rtc_alarm_callback alarm_callback;
+	void *alarm_user_data;
+#endif /* CONFIG_RTC_ALARM */
+#ifdef CONFIG_RTC_UPDATE
+	rtc_update_callback update_callback;
+	void *update_user_data;
+#endif /* CONFIG_RTC_UPDATE */
+#endif /* DS3231_INT1_GPIOS_IN_USE */
+};
+
+static int ds3231_read_regs(const struct device *dev, uint8_t addr, void *buf, size_t len)
```

Comment:
Does *buf need to be a void cast? It seems like everywhere it is used it is cast as `uint8_t`?